### DISCUSSION TITLE
v2.4.0

### DISCUSSION THREAD

**Discussion by alirezanet** [alirezanet/Gridify#40] [General]
- Added `QueryBuilder` class to help manual query creation process.
- Add Evaluation logic, Issue #38 
- Fix access modifier of some internal classes
- Expose public access to `PredicateBuilder`
- Expose public access to `SyntaxTree`

What is your suggestion to improve the new QueryBuilder class?

<hr /><em>This discussion was created from the release <a href='https://github.com/alirezanet/Gridify/releases/tag/v2.4.0'>v2.4.0</a>.</em>

**Comment by cirrusone** [comment]
Some other ideas...

1) Ability to validate filter strings without raising errors. The following raises an expected error `Gridify.GridifyMapperException HResult=0x80131500 Message=Mapping 'XYZ' not found` if any part of the filter string doesn't map.

```
var gridifyFilterQuery = new GridifyQuery() { Filter = "XYZ" };
var gridifyFilterExpression = gridifyFilterQuery.GetFilteringExpression(gridifyDataMapper).Compile(); // Error Mapping 'XYZ' not found`
```

It would be nice if we could validate the filter, similar to this

```
var gridifyFilterQuery = new GridifyQuery() { Filter = "XYZ" };
bool validFilter =  gridifyFilterQuery.IsValid(gridifyDataMapper);
```

2) Front end, something like [this](https://querybuilder.js.org/) but C#/Gridify centric

**Reply by alirezanet** [reply]
interesting, but we can suppress the mapping errors using `IgnoreNotMappedFields` configuration.

```c#
	var gMapper = new GridifyMapper<MyClass>(q => q.IgnoreNotMappedFields = true );
	var gQuery = new GridifyQuery() { Filter = "XYZ=123" };
	var gridifyFilterExpression = gQuery.GetFilteringExpression<MyClass>(gMapper).Compile(); // it ignores XYZ 
// -----------------
      // Query builder version
	var builder = new QueryBuilder<MyClass>()
		.ConfigureDefaultMapper(q => q.IgnoreNotMappedFields = true )
		.AddCondition("XYZ=123")
		.Build();
```

**Reply by alirezanet** [reply]
also, You can check specific map using GridifyMapper
```c#
gMapper.HasMap("XYZ") // false
```

**Reply by alirezanet** [reply]
also, check out #34, isn't this enough?

**Comment by cirrusone** [comment]
I think these are close but I see a few potential issues,

i) ...configuration.IgnoreNotMappedFields = true - this hides any potential problems, whilst good for us developers it could cause issues for end users where items are filtered but should not have due to failed fields.
ii) ...Mapper.HasMap("XYZ") - this works for the field column but may not pick up an numeric field having filter "XYZ=text". Also, how do you get the individual fields if the filter is something "XYZ1=1,XYZ2=2,XYZ3=3" etc?

This isn't urgent so is probably worth... seeing if other users need this

**Reply by alirezanet** [reply]
You were right, we have this feature in `v2.4.7`. 
thanks
